package zlog_ceph_proto;

option optimize_for = SPEED;

message WriteOp {
  required uint64 epoch = 1;
  required uint64 pos = 2;
  required bytes data = 3;
}

message FillOp {
  required uint64 epoch = 1;
  required uint64 pos = 2;
}

message ReadOp {
  required uint64 epoch = 1;
  required uint64 pos = 2;
}

message TrimOp {
  required uint64 epoch = 1;
  required uint64 pos = 2;
}

message SealOp {
  required uint64 epoch = 1;  
}

message MaxPositionOp {
  required uint64 epoch = 1;  
}

message GetProjectionRet {
  required uint64 epoch = 1;
  optional bytes out = 2;  
}

message MaxPositionRet {
  required uint64 pos = 1;
}

message SetProjectionOp {
  required uint64 epoch = 1;
  required bytes data = 2;
}

message GetProjectionOp {
  required uint64 epoch = 1;
  required bool latest = 2; 
}

message GetLatestProjectionOp {
  required uint64 epoch = 1;
}

message SetLatestProjectionOp {
  required uint64 epoch = 1;
}

message GetEpochOp {
  required uint64 epoch = 1;
}

message LogEntry {
  enum Flags {
    DEFAULT = 0;
    INVALIDATED = 1;
    TRIMMED = 2;
  }
  optional Flags flags = 1 [default = DEFAULT];
  optional bytes data = 2;
}
